enhancements/hybrid-helm-operator.md

```diff
@@ -0,0 +1,236 @@
+---
+title: enable-hybrid-helm-operators
+authors:
+  - "@fabianvf"
+  - "@varshaprasad96
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-05-10
+last-updated: 2021-05-12
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# enable-hybrid-helm-operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding
+to implement the design.  For instance, 
+ > 1. This requires exposing previously private resources which contain sensitive
+  information.  Can we do this? 
+
+## Summary
+
+Allow helm-operator users to extend their operators using Golang. We would accomplish
+this by exposing all the functionality that the helm-operator has via a library
+and scaffolding out code that is equivalent to the default behavior of the helm-operator.
+The user could then modify that code as they see fit, effectively turning their helm-based
+operator into a Golang one, but leaving a few utilities specific to helm-operators available
+for convenience.
+
+
+## Motivation
+
+Helm is quite limited compared to Ansible or Golang, the other two technologies that
+the Operator SDK currently supports. Once a helm-operator author reaches the limits
+of what they can accomplish with Helm and begins to need additional functionality, the
+only choice they have is to throw away what they have and duplicate it with another
+technology. Even then, maintaining a compatible API is non-trivial as the helm-operator
+provides functionality that the user would need to reimplement. Rather than starting over,
+a hybrid operator approach would allow them to implement advanced functionality in
+Golang while leaving the existing helm functionality in place.
+
+### Goals
+
+- Provide a library containing some of the helper logic for interacting with Helm and
+  controller-runtime
+- Scaffold out whatever code does not fit easily with the library
+- If the user changes nothing in the code, their operator should behave exactly as it 
+  did when it was a plain helm-operator.
+- The user should be able to scaffold APIs in either Golang or Helm
+- In the end, we are trying to keep the ease of use of Helm-based operators, while adding
+  the extensibility of Golang-based operators.
+
+### Non-Goals
+
+- We will not continue to manage the hybrid operator in any way, once it is converted
+  it is essentially a Go-based operator and so will not make use of a base image with
+  dependencies, automatic upgrades of helm versions, or any other conveniences provided
+  by the base image. The user will need to manage and upgrade their code by hand.
```

Comment:
do we need to explicitly state that when we mention `hybrid` operator, this is essentially a golang + helm operator. There are possibilities of automating helm charts with ansible roles, but that is something which will not be dealt here and could be an extension of current work in future.

Reply:
yeah I'll replace `hybrid` with `hybrid helm` or maybe even `hybrid helm and go`